homeassistant/components/eq3btsmart/climate.py

```diff
@@ -0,0 +1,390 @@
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+
+        return self._is_available
+
+    @property
+    def hvac_action(self) -> HVACAction | None:
+        """Return the current running hvac operation."""
+
+        if self._thermostat.status is None:
+            return None
+
+        if self._thermostat.status.operation_mode == OperationMode.OFF:
+            return HVACAction.OFF
+
+        if self._thermostat.status.valve == 0:
+            return HVACAction.IDLE
+
+        return HVACAction.HEATING
+
+    @property
+    def current_temperature(self) -> float | None:
+        """Return the current temperature."""
+
+        match self._eq3_config.current_temp_selector:
+            case CurrentTemperatureSelector.NOTHING:
+                return None
+            case CurrentTemperatureSelector.VALVE:
+                if self._thermostat.status is None:
+                    return None
+
+                return float(self._thermostat.status.valve_temperature)
+            case CurrentTemperatureSelector.UI:
+                return self._target_temperature
+            case CurrentTemperatureSelector.DEVICE:
+                if self._thermostat.status is None:
+                    return None
+
+                return float(self._thermostat.status.target_temperature.value)
+            case CurrentTemperatureSelector.ENTITY:
+                state = self.hass.states.get(self._eq3_config.external_temp_sensor)
+                if state is not None:
+                    try:
+                        return float(state.state)
+                    except ValueError:
+                        pass
+
+        return None
+
+    @property
+    def target_temperature(self) -> float | None:
+        """Return the target temperature."""
+
+        match self._eq3_config.target_temp_selector:
+            case TargetTemperatureSelector.TARGET:
+                return self._target_temperature
+            case TargetTemperatureSelector.LAST_REPORTED:
+                if self._thermostat.status is None:
+                    return None
+
+                return float(self._thermostat.status.target_temperature.value)
+
+    async def async_set_temperature(self, **kwargs: Any) -> None:
+        """Set new target temperature."""
+
+        if ATTR_HVAC_MODE in kwargs:
+            mode = kwargs.get(ATTR_HVAC_MODE)
+
+            if mode is None:
+                return
+
+            if mode == HVACMode.OFF:
+                raise ServiceValidationError(
+                    f"[{self._eq3_config.name}] Can't change HVAC mode to off while changing temperature",
+                )
+
+            await self.async_set_hvac_mode(mode)
+
+        temperature = kwargs.get(ATTR_TEMPERATURE)
+
+        if temperature is None:
+            return
+
+        temperature = round(temperature * 2) / 2
+        temperature = min(temperature, self.max_temp)
+        temperature = max(temperature, self.min_temp)
```

Comment:
what exactly are you doing this for?

Reply:
The thermostats can only be set to temperatures that are increments of 0.5, so we round the temperature to the nearest half and also check that it is inside the temperature limits set by the thermostat. I guess this could also be extracted into the library.